File: c++/pro-con/ch3/3.1-binsearch/maximize_average.cpp
#include<algorithm>
#include<cstdio>
using namespace std;

const int MAX_N=10000;
const int MAX_W=1000000;
const double INF=1000000000.0;

int n,k;
int w[MAX_N], v[MAX_N];

double y[MAX_N]; // v - x*w

bool C(double x){
  for (int i=0; i<n; i++){
    y[i] = v[i] - x * w[i];
  }
  sort(y, y+n);
  double sum = 0;
  for (int i=0; i< k; i++){
    sum += y[n - i - 1];
  }
  return sum>=0;
}

void solve(){
  double lb=0, ub=INF;
  for(int i=0; i<100; i++){
    double mid=(lb+ub)/2;
    if(C(mid)) lb = mid;
    else ub = mid;
  }

  printf("%.2f\n", ub);
}

File: python/pro-con-python/ch3/sec1_binarySearch/MaximizeAverage.py
'''
Created on 2016/12/17

@author: kazuyoshi
'''
INF=1<<32

def C(V,W,x,k):
    y=[]
    for v,w in zip(V,W):
        y.append(v - x *w)
    y.sort(reverse=True)
    return sum(y[0:k])>=0

def solve(V,W,k):
    lb=0
    ub=INF
    for i in range(100):
        mid = (lb+ub)/2
        if C(V,W,mid,k):
            lb = mid
        else:
            ub=mid
    return ub

if __name__ == '__main__':
    n=3
    k=2
    W =[2,5,2]
    V =[2,3,1]
    print("{:.2f}".format(solve(V,W,k)))